### ISSUE TITLE
[BUG] dilateActiveValues() function creates Artifacts

### ISSUE THREAD

**Issue by mifth** [AcademySoftwareFoundation/openvdb#1078]
### Environment
**Operating System:** Windows 10
**Version / Commit SHA:** May 21, 2021, 316da7944c7c289306a00d2badeec39ea3fdd7d1
**Other:** Cmake

### Describe the bug
I found that **openvdb::tools:dilateValues(tree, iterations)** is deprecated and decided to test the new **dilateActiveValues()** function. And I found that **dilateActiveValues()**  makes artifacts if I will use gaussian filter after the function.
Here is my comparison of old function and the new one:
![image](https://user-images.githubusercontent.com/2251298/120070699-bc758800-c083-11eb-932a-b096b51a1693.png)

### To Reproduce
Load my vdb file: 
[vdb_out_1.zip](https://github.com/AcademySoftwareFoundation/openvdb/files/6564685/vdb_out_1.zip)

Then run code like this:
```
// load a grid from the vdb file

// openvdb::tools::dilateVoxels(grid->tree(), 12);    // OK
openvdb::tools::dilateActiveValues(grid->tree(), 12);   // ARTIFACTS

// apply a gaussian filter
openvdb::tools::Filter<openvdb::FloatGrid> lsf(*grid);
lsf.setGrainSize(1);
lsf.gaussian(2, 3);

// convert to mesh
openvdb::tools::VolumeToMesh mesher(0.0, 0.0);
mesher.operator()<openvdb::FloatGrid>( grid.operator*() );
```

### Expected behavior
No artifacts

**Comment by Idclip** [comment]
Hey @mifth, thanks a lot for reporting. This is a guess, but I suspect it's because the new version of the dilate function produces a more sparse result which the level set filter can't handle. There is an enum you can pass in to dilateActiveValues, can you try calling it with the EXPAND_TILES variant and let me know how that goes. I can further explain what's happening here should that fix it.

**Comment by mifth** [comment]
Hi @Idclip , thank you for fast reply. 
openvdb::tools::EXPAND_TILES and openvdb::tools::IGNORE_TILES work nicely and fix the issue.

May I suggest to set **mode = openvdb::tools::EXPAND_TILES** by default so that it worked as in old version by default?

**Comment by mifth** [comment]
And another question:
Does **openvdb::tools::erodeActiveValues()** function works as an old version? I mean in the same way?

**Comment by Idclip** [comment]
The TL;DR is that, to achieve the **exact** same output of `tools::dilateVoxel` and `tools::erodeVoxels`, you should use the `tools::IGNORE_TILES` with both `tools::dilateActiveValues` and `tools::erodeActiveValues`. There is one further difference with `tools::erodeActiveValues` in that the old method performed a final call to `tools::pruneLevelSet` (regardless of the input tree type). I will update the deprecation messages to be clearer.

-----------

I'll go into more detail as there's a few things to unpack. Should your input tree contain tiles (that is, any node in the hierarchy that contains an active value and is **not** a `LeafNode`), the above techniques will not produce a true dilation/erosion. A further issue with the old methods is that they do not preserve the sparsity of the tree. `tools::PRESERVE_TILES` will do this, ensuring that any newly created and constant dense `LeafNodes` are appropriately collapsed. 

Ideally every tool which modifies a tree should try to preserve its sparse structure and every tool which runs over a tree should be able to support any tree configuration/node structure. Neither were true for the old methods. Unfortunately, the latter was also not true for the `tools::Filter<>` class. What you're observing here is ultimately a problem with the default behaviour `tools::Filter<>`.

Unfortunately, unlike the morphology methods where we've added new API, the `tools::Filter<>` class has no such wrappers. Users have to invoke instances of the class directly. We can't change the default behaviour of this object (even if the results are more correct). To that end, we extended `tools::Filter<>` to handle tiles, but this behaviour is off by default. You can turn it on with `tools::Filter<>::setProcessTiles(true);`. In your example, you would do:

```
openvdb::tools::Filter<openvdb::FloatGrid> lsf(*grid);
lsf.setProcessTiles(true);
lsf.setGrainSize(1);
lsf.gaussian(2, 3);
```

This will allow you to maintain the default arguments of `tools::dilateActiveValues` and handle the "better" (more sparse) tree result during filtering.

So that leaves us with two routes. 1)

```
openvdb::tools::dilateActiveValues(grid->tree(), 12, openvdb::tools::NN_FACE, openvdb::tools::IGNORE_TILES); 
openvdb::tools::Filter<openvdb::FloatGrid> lsf(*grid);
lsf.setGrainSize(1);
lsf.gaussian(2, 3);
```

or 2)

```
openvdb::tools::dilateActiveValues(grid->tree(), 12); // defaults to openvdb::tools::PRESERVE_TILES
openvdb::tools::Filter<openvdb::FloatGrid> lsf(*grid);
lsf.setProcessTiles(true); 
lsf.setGrainSize(1);
lsf.gaussian(2, 3);
```

The answer to "which is better?" can depend, but it'll generally be (2) and you'll start to see the advantages as your kernel/input size grow. This is because:

 - The more dilation steps you do, the more dense topology you are creating which will typically have uniform values. This topology can be collapsed which can lower peak memory usage and overall tree footprint.
 - The `tools::Filter<>` class will get progressively slower with dense topology. With tile support, the Filter class will detect which tiles it **needs** to filter by checking neighbouring topology and **only** densify them when required.

The main reason to use (1) would be if you know your input contains no tiles **and** need to keep your topology dense to use other tools which still don't support tiles.

**Comment by mifth** [comment]
Thank you!
